lightly/utils/benchmarking.py

```diff
@@ -171,59 +174,61 @@ def __init__(
         self.knn_k = knn_k
         self.knn_t = knn_t
 
-        # create dummy param to keep track of the device the model is using
-        self.dummy_param = nn.Parameter(torch.empty(0))
+        self._train_features: Optional[Tensor] = None
+        self._train_targets: Optional[Tensor] = None
+        self._val_predicted_labels: List[Tensor] = []
+        self._val_targets: List[Tensor] = []
 
-    def training_epoch_end(self, outputs):
-        # update feature bank at the end of each training epoch
-        self.backbone.eval()
-        self.feature_bank = []
-        self.targets_bank = []
+    def on_validation_epoch_start(self) -> None:
+        train_features = []
+        train_targets = []
         with torch.no_grad():
             for data in self.dataloader_kNN:
                 img, target, _ = data
-                img = img.to(self.dummy_param.device)
-                target = target.to(self.dummy_param.device)
+                img = img.to(self.device)
+                target = target.to(self.device)
                 feature = self.backbone(img).squeeze()
                 feature = F.normalize(feature, dim=1)
-                self.feature_bank.append(feature)
-                self.targets_bank.append(target)
-        self.feature_bank = torch.cat(self.feature_bank, dim=0).t().contiguous()
-        self.targets_bank = torch.cat(self.targets_bank, dim=0).t().contiguous()
-        self.backbone.train()
-
-    def validation_step(self, batch, batch_idx):
+                if dist.is_initialized() and dist.get_world_size() > 0:
+                    # gather features and targets from all processes
+                    feature = torch.cat(dist.gather(feature), 0)
+                    target = torch.cat(dist.gather(target), 0)
+                train_features.append(feature)
+                train_targets.append(target)
+        self._train_features = torch.cat(train_features, dim=0).t().contiguous()
+        self._train_targets = torch.cat(train_targets, dim=0).t().contiguous()
+
+    def validation_step(self, batch, batch_idx) -> None:
         # we can only do kNN predictions once we have a feature bank
-        if hasattr(self, "feature_bank") and hasattr(self, "targets_bank"):
+        if self._train_features is not None and self._train_targets is not None:
             images, targets, _ = batch
             feature = self.backbone(images).squeeze()
             feature = F.normalize(feature, dim=1)
-            pred_labels = knn_predict(
+            predicted_labels = knn_predict(
                 feature,
-                self.feature_bank,
-                self.targets_bank,
+                self._train_features,
+                self._train_targets,
                 self.num_classes,
                 self.knn_k,
                 self.knn_t,
             )
-            num = images.size()
-            top1 = (pred_labels[:, 0] == targets).float().sum()
-            return (num, top1)
```

Comment:
Why is it moved instead into the `on_validation_epoch_end`?

Reply:
> Why is it moved instead into the on_validation_epoch_end?

Because PyTorch Lightning 2.0 removed the `training_epoch_end ` and `validation_epoch_end ` methods.

> Why is nothing returned anymore?

Previously, PyTorch Lightning would aggregate all returned values from `validation_step` and pass them to `validation_epoch_end` as the `outputs` argument. This was removed in 2.0 and now it is up to users to keep track of the values.